crates/orchestrator/src/config.rs

```diff
@@ -78,6 +94,151 @@ pub struct NetworkConfig<KEY: SignatureKey, ELECTIONCONFIG: ElectionConfig> {
     pub da_web_server_config: Option<WebServerConfig>,
 }
 
+pub enum NetworkConfigSource {
+    Orchestrator,
+    File,
+}
+
+impl<K: SignatureKey, E: ElectionConfig> NetworkConfig<K, E> {
+    /// Asynchronously retrieves a `NetworkConfig` either from a file or from an orchestrator.
+    ///
+    /// This function takes an `OrchestratorClient`, an identity string, and an optional file path.
+    ///
+    /// If a file path is provided, the function will first attempt to load the `NetworkConfig` from the file.
+    /// If the file does not exist or cannot be read, the function will fall back to retrieving the `NetworkConfig` from the orchestrator.
+    /// The retrieved `NetworkConfig` is then saved back to the file for future use.
+    ///
+    /// If no file path is provided, the function will directly retrieve the `NetworkConfig` from the orchestrator.
+    ///
+    /// # Arguments
+    ///
+    /// * `client` - An `OrchestratorClient` used to retrieve the `NetworkConfig` from the orchestrator.
+    /// * `identity` - A string representing the identity for which to retrieve the `NetworkConfig`.
+    /// * `file` - An optional string representing the path to the file from which to load the `NetworkConfig`.
+    ///
+    /// # Returns
+    ///
+    /// This function returns a tuple containing a `NetworkConfig` and a `NetworkConfigSource`. The `NetworkConfigSource` indicates whether the `NetworkConfig` was loaded from a file or retrieved from the orchestrator.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// let client = OrchestratorClient::new();
+    /// let identity = "my_identity".to_string();
+    /// let file = Some("/path/to/my/config".to_string());
+    /// let (config, source) = NetworkConfig::from_file_or_orchestrator(client, file).await;
+    /// ```
+    pub async fn from_file_or_orchestrator(
+        client: &OrchestratorClient,
+        file: Option<String>,
+    ) -> (NetworkConfig<K, E>, NetworkConfigSource) {
+        if let Some(file) = file {
+            // if we pass in file, try there first
+            match Self::from_file(file.clone()) {
+                Ok(config) => (config, NetworkConfigSource::File),
+                Err(e) => {
+                    // fallback to orchestrator
+                    error!("{e}");
+
+                    let config = client.get_config(client.identity.clone()).await;
+
+                    // save to file if we fell back
+                    if let Err(e) = config.to_file(file) {
+                        error!("{e}");
```

Comment:
Should we handle this error either here or in the caller? I think we should do more than just the logging.

Reply:
Hmm, I think maybe we should consider leaving it here. This function is supposed to be higher-level for the common use case of "load from either", but if the caller wants to add additional functionality they can call the functions directly and not use this one

Reply:
Makes sense! Logging the error should suffice then.